### ISSUE TITLE
openfoam: broken modules (with zsh)

### ISSUE THREAD

**Issue by davydden** [spack/spack#13504]
Looks like something is wrong with modules from `openfoam`. If I generate module files, spack shows an error for `openfoam` package:

```
==> Warning: Quotes in command arguments can confuse scripts like configure.
  The following arguments may cause problems when executed:
      source /spack/opt/spack/linux-centos7-skylake_avx512/gcc-9.2.0/openfoam-1906-gf5sdxweeqsj2awgyc5bu2uy5grudvmf/etc/bashrc &> /dev/null && python -c "import os, json; print(json.dumps(dict(os.environ)))"
  Quotes aren't needed because spack doesn't use a shell.
  Consider removing them
==> OpenFOAM bashrc env: /spack/opt/spack/linux-centos7-skylake_avx512/gcc-9.2.0/openfoam-1906-gf5sdxweeqsj2awgyc5bu2uy5grudvmf/etc/bashrc
==> Warning: Quotes in command arguments can confuse scripts like configure.
  The following arguments may cause problems when executed:
      source /spack/opt/spack/linux-centos7-skylake_avx512/gcc-9.2.0/openfoam-1906-gf5sdxweeqsj2awgyc5bu2uy5grudvmf/etc/bashrc &> /dev/null && python -c "import os, json; print(json.dumps(dict(os.environ)))"
  Quotes aren't needed because spack doesn't use a shell.
  Consider removing them
==> OpenFOAM bashrc env: /spack/opt/spack/linux-centos7-skylake_avx512/gcc-9.2.0/openfoam-1906-gf5sdxweeqsj2awgyc5bu2uy5grudvmf/etc/bashrc
```

when loading a module, I get from `lmod`:
```
invalid command
name ""
```
or from `environment modules`:
```
Module ERROR: invalid command name ""
```

This has to do with 
```
prepend-path BASH_FUNC__module_raw() "() {  unset _mlshdbg;
...blahblahblah...
}"
```
If I manually delete it, I can at least load modules and get environment variables set correctly.

**Comment by davydden** [comment]
@olesenm do you happen to have an idea what could be wrong with openfoam modules?

**Comment by olesenm** [comment]
@alalazo is the guru for these things.

**Comment by davydden** [comment]
@alalazo if/when you have a moment, could you please have a look at this module thing issue? The problem persist even with latest 0.14.2:

```
spack --stacktrace load openfoam 
lib/spack/spack/util/executable.py:152 ==> Warning: Quotes in command arguments can confuse scripts like configure.
  The following arguments may cause problems when executed:
      source /dev/null &> /dev/null && python -c "import os, json; print(json.dumps(dict(os.environ)))"
  Quotes aren't needed because spack doesn't use a shell.
  Consider removing them
lib/spack/spack/util/executable.py:152 ==> Warning: Quotes in command arguments can confuse scripts like configure.
  The following arguments may cause problems when executed:
      source /spack/opt/spack/linux-centos7-skylake_avx512/gcc-9.2.0/openfoam-1912_200403-dpmvivsqssys6kbqhzvjklroym3mjpyp/etc/bashrc &> /dev/null && python -c "import os, json; print(json.dumps(dict(os.environ)))"
  Quotes aren't needed because spack doesn't use a shell.
  Consider removing them
spack:140: no such file or directory: lib/spack/spack/user_environment.py:89
```

This is no longer related to `BASH_FUNC__module_raw` (the function is not even there).

A workaround is to use modules manually:  `module load openfoam-1912_200403-gcc-9.2.0-dpmvivs`.

**Comment by becker33** [comment]
@davydden @olesenm I think this is still related to `BASH_FUNC__module_raw` coming from the bashrc file that is sourced in `setup_run_environment`. I think if you add `BASH_FUNC_*` to the blacklist in that method I think it should resolve it.

I have not installed openfoam locally to test as I'm hoping someone can test my theory without having to spend the time to install openfoam, but I can try it out if you don't have an installation handy to test it out on.

**Comment by olesenm** [comment]
Hi @becker33 

I have used `grep -i BASH` on all openfoam module files generated by spack on my installation (openfoam 1806,1812,1906,1912) and I don't see any instances at all.

if you want to test a minimal (non-functional) openfoam installation without the long compilation times, there is a backdoor that I use. It could be more elegant, but is really only ever meant as a diagnostics hack.

Change the `packages/openfoam/package.py` around line 342 to include the `common/spack-dummy-Allwmake':
```
    # General patches
    common = ['spack-Allwmake', 'README-spack', 'spack-dummy-Allwmake']
```

And later around line 372, specify that it should be used:
```
    build_script = './spack-dummy-Allwmake'
```

This builds the equivalent of "hello world" with a single application and a single library. It builds in a a few seconds and satisfies the rest of the build and installation process, which makes it perfect for testing the rest of the installation logic without waiting around (ie, good for quickly testing your hypothesis and hammering out details).

Hope that helps,

/mark

**Comment by xdelaruelle** [comment]
@alalazo @becker33 I was able to reproduce the `spack:140: no such file or directory: lib/spack/spack/user_environment.py:89` issue on Spack 0.14.2 as described above. But I did not achieve to generate on this version of Spack an openfoam modulefile which includes the `prepend-path BASH_FUNC__module_raw()` command.

On current Spack develop `spack load openfoam` works fine. No more warning or error message.

So from what I observe with my basic tests, there is no more issue nowadays either on modulefile generation or `spack load` sides.

What could be done however is to add `BASH_FUNC__module_raw()` to the environment variable exclusion list in `lib/spack/spack/util/environment.py`. `_module_raw()` is the shell function coming together with `module()` on Environment Modules to redirect output from stderr to stdout. I will elaborate a pull request for that.
